library/common/http/client.h

```diff
@@ -135,16 +141,53 @@ class Client : public Logger::Loggable<Logger::Id::http> {
       NOT_IMPLEMENTED_GCOVR_EXCL_LINE;
     }
     bool streamErrorOnInvalidHttpMessage() const override { return false; }
+
     void encodeMetadata(const MetadataMapVector&) override { NOT_IMPLEMENTED_GCOVR_EXCL_LINE; }
 
+    void onHasBufferedData() { direct_stream_.runHighWatermarkCallbacks(); }
+    void onBufferedDataDrained() { direct_stream_.runLowWatermarkCallbacks(); }
+
+    // To be called by mobile library when explicitly buffering and more data is wanted.
+    // If bytes are available, the bytes available (up to the limit of
+    // bytes_to_send) will be shipped the bridge immediately.
+    //
+    // If no bytes are available, the next time data is received from the
+    // network, up to bytes_to_send bytes will be shipped to the bridge.
+    //
+    // Bytes will only be sent up once, even if the bytes available are fewer
+    // than bytes_to_send.
+    void resumeData(int32_t bytes_to_send);
+
+    bool explicitBuffering() const { return explicit_buffering_; }
+
   private:
+    bool hasBufferedData() { return response_data_.get() && response_data_->length() != 0; }
+
+    void setExplicitlyBuffering() { explicit_buffering_ = true; }
+    void sendDataToBridge(Buffer::Instance& data, bool end_stream);
+    void sendTrailersToBridge(const ResponseTrailerMap& trailers);
+
     DirectStream& direct_stream_;
     const envoy_http_callbacks bridge_callbacks_;
     Client& http_client_;
     absl::optional<envoy_error_code_t> error_code_;
     absl::optional<envoy_data> error_message_;
     absl::optional<int32_t> error_attempt_count_;
     bool success_{};
+    // Buffered response data when explicitly buffering.
+    Buffer::InstancePtr response_data_;
+    ResponseTrailerMapPtr response_trailers_;
+    // True if the bridge should operate in explicitly buffering mode, and only send
+    // data when it is requested by the caller.
+    bool explicit_buffering_{};
+    bool response_headers_sent_{};
+    // Called in closeStream() to communicate that the end of the stream has
+    // been received by the DirectStreamCallbacks.
+    bool end_stream_read_{};
+    // Set true when the end stream has been communicated up the bridge.
+    bool end_stream_communicated_{};
```

Comment:
nit:
`response_headers_forwarded_`
`end_stream_received_`
`end_stream_forwarded_`

would align slightly better with terminology we've used elsewhere.